### PR TITLE
Wait for replicas

### PR DESCRIPTION
**Pull request by sarahbx**
##### Short description:
Add additional conditions for break

### CODE REVIEW

File: ocp_resources/deployment.py [hunk 1]
```diff
@@ -32,13 +32,13 @@ def scale_replicas(self, replica_count=int):
         LOGGER.info(f"Set deployment replicas: {replica_count}")
         return self.update(resource_dict=body)
 
-    def wait_for_replicas(self, deployed=True, timeout=TIMEOUT):
+    def wait_for_replicas(self, timeout=TIMEOUT, deployed=True):
```
Comment (myakove) [inline ocp_resources/deployment.py:35]: Please do not switch places of the args, this may break existing code

File: ocp_resources/deployment.py [hunk 2]
```diff
@@ -32,13 +32,13 @@ def scale_replicas(self, replica_count=int):
...
         """
         Wait until all replicas are updated.
 
         Args:
-            deployed (bool): True for replicas deployed, False for no replicas.
             timeout (int): Time to wait for the deployment.
+            deployed (bool): (Deprecated) True for replicas deployed, False for no replicas.  # TODO: Deprecated
```
Comment (myakove) [inline ocp_resources/deployment.py:41]: Add Deprecated warrning?

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/deployment.py [hunk 3]
```diff
@@ -52,11 +52,8 @@ def wait_for_replicas(self, deployed=True, timeout=TIMEOUT):
...
-                status = sample.items[0].status
-                if deployed:
-                    if status.replicas == status.availableReplicas:
-                        return
-                else:
-                    if not status.availableReplicas:
-                        return
+            instance = sample.items[0] if sample.items else None
+            if instance and instance.spec.replicas == (
+                instance.status.replicas or 0
```
Comment (rnetser) [inline ocp_resources/deployment.py:57]: if `instance.spec.replicas` is 0, `instance.status.replicas` does not exist:

```
status:
  conditions:
  - lastTransitionTime: "2021-12-05T11:23:31Z"
    lastUpdateTime: "2021-12-05T11:24:01Z"
    message: ReplicaSet "ssp-operator-58466c6fb9" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  - lastTransitionTime: "2021-12-11T19:09:02Z"
    lastUpdateTime: "2021-12-11T19:09:02Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  observedGeneration: 36

```

File: ocp_resources/deployment.py (hunk 2 above)
Comment (rnetser) [reply ocp_resources/deployment.py:41]: I do not think we should remove this; a user should request an expected status.

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/deployment.py [hunk 4]
```diff
@@ -52,11 +52,17 @@ def wait_for_replicas(self, deployed=True, timeout=TIMEOUT):
...
+                spec_replicas = instance.spec.replicas
+                total_replicas = instance.status.replicas or 0
+                updated_replicas = instance.status.updatedReplicas or 0
+                available_replicas = instance.status.availableReplicas or 0
+
+                if (
+                    deployed
+                    and spec_replicas
+                    and updated_replicas >= spec_replicas
+                    and available_replicas >= spec_replicas
```
Comment (rnetser) [inline ocp_resources/deployment.py:66]: ditto

**Review by rnetser — CHANGES_REQUESTED** [review]

**Review by myakove — APPROVED** [review]

**Review by rnetser — APPROVED** [review]

**Pull request merged**